pyphare/pyphare/core/operators.py

```diff
@@ -0,0 +1,150 @@
+import numpy as np
+# from pyphare.pharesee.hierarchy import PatchLevel, Patch, FieldData
+from pyphare.pharesee.hierarchy import ScalarField, VectorField  # TensorField
+from pyphare.pharesee.hierarchy import compute_hier_from
+from pyphare.pharesee.hierarchy import rename
+
+
+def _compute_dot_product(patch_datas, **kwargs):
+
+    ref_name = next(iter(patch_datas.keys()))
+
+    dset = patch_datas["left_x"].dataset[:] * patch_datas["right_x"].dataset[:]\
+         + patch_datas["left_y"].dataset[:] * patch_datas["right_y"].dataset[:]\
+         + patch_datas["left_z"].dataset[:] * patch_datas["right_z"].dataset[:]
+
+    return ({"name": 'scalar', "data": dset, "centering": patch_datas[ref_name].centerings},)
```

Comment:
this (line 16) looks like it returns a tuple of size one, is that the case? and if so, is that required?

Reply:
Yes, if I'm correct, the `compute` call back returns a set of dict containing the `name`, `data` and `centering`. At the end of the game, this is `new_patchdatas_from` which call the `compute callback, and put the results in `datas`. As we iterate in the elements of `datas`, I think it has to be a tuple.